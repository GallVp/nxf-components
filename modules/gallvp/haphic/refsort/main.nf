process HAPHIC_REFSORT {
    tag "$meta.id"
    label 'process_single'

    container 'docker.io/gallvp/haphic:1.0.7--hdfd78af_0'

    input:
    tuple val(meta), path(agp)
    path paf
    path fasta
    val ref_order

    output:
    tuple val(meta), path('*.agp')      , emit: agp
    tuple val(meta), path('*.fasta')    , emit: fasta   , optional: true
    path "versions.yml"                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def fasta_arg = fasta ? "--fasta $fasta" : ''
    def ref_order_arg = ref_order ? "--ref_order $ref_order" : ''
    def fout_arg = fasta ? "--fout ${prefix}.fasta" : ''
    if( "$agp" == "${prefix}.agp" ) error "Input and output AGP names are the same, use \"task.ext.prefix\" to disambiguate!"
    if( "$fasta" == "${prefix}.fasta" ) error "Input and output Fasta names are the same, use \"task.ext.prefix\" to disambiguate!"
    """
    haphic \\
        refsort \\
        $args \\
        $ref_order_arg \\
        $fasta_arg \\
        $fout_arg \\
        $agp \\
        $paf \\
        > ${prefix}.agp

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        haphic: \$(haphic --version |& sed -n 's|HapHiC \\(.*\\)|\\1|p')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    if( "$agp" == "${prefix}.agp" ) error "Input and output AGP names are the same, use \"task.ext.prefix\" to disambiguate!"
    if( "$fasta" == "${prefix}.fasta" ) error "Input and output Fasta names are the same, use \"task.ext.prefix\" to disambiguate!"
    """
    touch ${prefix}.agp

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        haphic: \$(haphic --version |& sed -n 's|HapHiC \\(.*\\)|\\1|p')
    END_VERSIONS
    """
}
