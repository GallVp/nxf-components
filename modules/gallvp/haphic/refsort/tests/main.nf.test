nextflow_process {

    name "Test Process HAPHIC_REFSORT"
    script "../main.nf"
    config "./nextflow.config"
    process "HAPHIC_REFSORT"

    tag "modules"
    tag "modules_gallvp"
    tag "haphic"
    tag "haphic/refsort"
    tag "minimap2/align"
    tag "gunzip"

    setup {
        run("MINIMAP2_ALIGN") {
            script "../../../minimap2/align/main.nf"
            process {
                """
                input[0] = [
                    [ id:'test', single_end:true ], // meta map
                    file('https://github.com/Plant-Food-Research-Open/assemblyqc/raw/4f9fba392cf7212d61c6245fd72339e09164ce44/tests/hic/testdata/HYh1h2_Chr01_6000000_6015000.fasta.gz', checkIfExists: true)
                ]
                input[1] = [
                    [ id:'test_ref' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/eukaryotes/actinidia_chinensis/genome/chr1/genome.fasta.gz', checkIfExists: true)
                ]
                input[2] = false
                input[3] = []
                input[4] = false
                input[5] = false
                """
            }
        }
    }

    test("hongyang - hic - fasta") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file('https://github.com/Plant-Food-Research-Open/assemblyqc/raw/4f9fba392cf7212d61c6245fd72339e09164ce44/tests/hic/testdata/HYh1h2_Chr01_6000000_6015000.agp', checkIfExists: true),
                ]
                input[1] = MINIMAP2_ALIGN.out.paf.map { it[1] }
                input[2] = []
                input[3] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("hongyang - hic - fasta - fout") {

        setup {
            run("GUNZIP") {
                script "../../../gunzip/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test', single_end:true ], // meta map
                        file('https://github.com/Plant-Food-Research-Open/assemblyqc/raw/4f9fba392cf7212d61c6245fd72339e09164ce44/tests/hic/testdata/HYh1h2_Chr01_6000000_6015000.fasta.gz', checkIfExists: true)
                    ]
                    """
                }
            }
        }

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file('https://github.com/Plant-Food-Research-Open/assemblyqc/raw/4f9fba392cf7212d61c6245fd72339e09164ce44/tests/hic/testdata/HYh1h2_Chr01_6000000_6015000.agp', checkIfExists: true),
                ]
                input[1] = MINIMAP2_ALIGN.out.paf.map { it[1] }
                input[2] = GUNZIP.out.gunzip.map { it[1] }
                input[3] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}