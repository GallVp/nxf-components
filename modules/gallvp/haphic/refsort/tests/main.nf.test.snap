{
    "hongyang - hic - fasta - fout": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.agp:md5,23845f43c5a8aa2ad5a1c25660831eba"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.fasta:md5,bcc14495671d77033b36be11fc20652c"
                    ]
                ],
                "2": [
                    "versions.yml:md5,7a386c0c393a7ffdbc237e422e32bd59"
                ],
                "agp": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.agp:md5,23845f43c5a8aa2ad5a1c25660831eba"
                    ]
                ],
                "fasta": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.fasta:md5,bcc14495671d77033b36be11fc20652c"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,7a386c0c393a7ffdbc237e422e32bd59"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-19T10:01:30.647079"
    },
    "hongyang - hic - fasta": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.agp:md5,23845f43c5a8aa2ad5a1c25660831eba"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,7a386c0c393a7ffdbc237e422e32bd59"
                ],
                "agp": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.agp:md5,23845f43c5a8aa2ad5a1c25660831eba"
                    ]
                ],
                "fasta": [
                    
                ],
                "versions": [
                    "versions.yml:md5,7a386c0c393a7ffdbc237e422e32bd59"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-19T10:01:20.254858"
    }
}